File: frontend/src/types/agent-decision.ts
/** Action types an agent can perform */
export type ActionType =
  // Cooperative
  | 'move' | 'gather' | 'repair' | 'trade' | 'talk'
  | 'vote' | 'rest' | 'observe' | 'heal'
  // Selfish
  | 'hoard' | 'sabotage' | 'explore' | 'accuse' | 'steal' | 'scheme'
  // Aggressive
  | 'attack' | 'threaten' | 'stab' | 'shoot' | 'poison'
  // Social / expressive
  | 'dance' | 'pray' | 'rally' | 'mourn' | 'celebrate' | 'argue'
  // Biological (dark humor)
  | 'pee' | 'poop' | 'vomit' | 'sleep' | 'eat' | 'drink'
  // Meta
  | 'investigate' | 'monologue' | 'panic' | 'breakdown'

/** Cooperation intent level */
export type CooperationIntent = 'high' | 'medium' | 'low' | 'none'

/** An action chosen by an agent */
export interface AgentAction {
  type: ActionType
  target?: string
  location?: string
}

/** Dialogue from an agent to another */
export interface AgentDialogue {
  target: string
  message: string
}

/** Structured decision output from an agent's LLM (matches shared/schemas/agent_decision.json) */
export interface AgentDecision {
  innerThought: string
  suspicion: string | null
  action: AgentAction
  dialogue: AgentDialogue | null
  goalProgress: string
  cooperationIntent: CooperationIntent
}

File: backend/app/schemas/agent_decision.py
from typing import Literal, cast

from pydantic import Field

from app.actions import DECISION_ACTION_IDS
from app.schemas.common import APIModel

DecisionActionType = Literal[
    "move",
    "gather",
    "repair",
    "trade",
    "talk",
    "hoard",
    "sabotage",
    "explore",
    "accuse",
    "vote",
    "rest",
    "observe",
    "attack",
    "threaten",
    "stab",
    "shoot",
    "poison",
    "dance",
    "pray",
    "rally",
    "mourn",
    "celebrate",
    "argue",
    "pee",
    "poop",
    "vomit",
    "sleep",
    "eat",
    "drink",
    "investigate",
    "monologue",
    "panic",
    "breakdown",
    "self_sacrifice",
]
DECISION_ACTION_TYPES: tuple[DecisionActionType, ...] = cast(
    tuple[DecisionActionType, ...], DECISION_ACTION_IDS
)
CooperationIntent = Literal["high", "medium", "low", "none"]

AGENT_INNER_THOUGHT_MAX_LENGTH = 160
AGENT_DECISION_MAX_TOKENS = 384


class DecisionAction(APIModel):
    type: DecisionActionType
    target: str | None = None
    location: str | None = None


class Dialogue(APIModel):
    target: str
    message: str


class AgentDecision(APIModel):
    inner_thought: str = Field(
        min_length=1,
        max_length=AGENT_INNER_THOUGHT_MAX_LENGTH,
        description=(
            "A brief window into the agent's immediate reasoning. Keep it to 1-2 short "
            "sentences and avoid monologues."
        ),
    )
    suspicion: str | None = None
    action: DecisionAction
    dialogue: Dialogue | None = None
    goal_progress: str
    cooperation_intent: CooperationIntent
